### ISSUE TITLE
Cancelling causes a crash

### ISSUE THREAD

**Issue by Geocfu** [d4rken/sdmaid-public#323]
While the operation has started, if i press the cancel button, the app will force close

**Comment by d4rken** [comment]
A specific tool?

**Comment by Geocfu** [comment]
I think it does not matter. It has happend on corpse finder and on systemcleaner.

**Comment by d4rken** [comment]
I'm seeing some related crashes in the bug tracker, can you give me your install ID?
It's under Overview > SD Maid, you can tap it to copy the ID to clipboard.
With it i can find the crash report from your device.

**Comment by Geocfu** [comment]
721382ad-3697-49fd-8416-cec333ca6ac2

**Comment by d4rken** [comment]
Hm weird, only a crash from 4 days ago.

I think i can reproduce it though, not sure if its the same crash but it's when canceling.

**Comment by Geocfu** [comment]
Do you want me to take a logcat? Just tell me how.

**Comment by d4rken** [comment]
I think doing a debug run while it crashes actually gets the crash own SD Maids own log, that would be cool to try :sunglasses: .

**Comment by Geocfu** [comment]
Ok i did it twice. So we have a log with no force close while canceling and a log with force close due to canceling. Take a look.

[sdmaid_logfile_1457365331982NOCRASH.txt](https://github.com/d4rken/sdmaid-public/files/161905/sdmaid_logfile_1457365331982NOCRASH.txt)
[sdmaid_logfile_1457365488554CRASHED.txt](https://github.com/d4rken/sdmaid-public/files/161906/sdmaid_logfile_1457365488554CRASHED.txt)

**Comment by d4rken** [comment]
Crash that takes SD Maid doesn't seem to be logged by SD Maid itself, which would explain why I get no automatic crash report from your for that. Not sure why that is, likely doesn't deal well with multithreading.

You could do a normal logcat with an external tool, if you don't that's okay too. I'm pretty sure it only happens if you cancel while a shell using root is open and SD Maid then tries to cancel the root shell.
It tries to get the PID of that shell and crashes while parsing for the PID.

**Comment by Geocfu** [comment]
Did i got it correct?
[locat.txt](https://github.com/d4rken/sdmaid-public/files/161975/locat.txt)

**Comment by d4rken** [comment]
Although the file is weirdly formated (open it and see), yes you did the crash is on there :+1: 

```
E/AndroidRuntime( 8271): FATAL EXCEPTION: Thread-3144
E/AndroidRuntime( 8271): Process: eu.thedarken.sdm, PID: 8271
E/AndroidRuntime( 8271): java.lang.UnsupportedOperationException
E/AndroidRuntime( 8271):        at java.util.AbstractList.add(AbstractList.java:404)
E/AndroidRuntime( 8271):        at java.util.AbstractList.add(AbstractList.java:425)
E/AndroidRuntime( 8271):        at eu.thedarken.sdm.tools.n.a(PIDKiller.java:66)
E/AndroidRuntime( 8271):        at eu.thedarken.sdm.tools.n.a(PIDKiller.java:103)
E/AndroidRuntime( 8271):        at eu.thedarken.sdm.tools.shell.a.a.i(InteractiveShell.java:415)
E/AndroidRuntime( 8271):        at eu.thedarken.sdm.tools.io.hybrid.d.d.i(HybridReaderStreamer.java:90)
E/AndroidRuntime( 8271):        at eu.thedarken.sdm.systemcleaner.SystemCleanerWorker.a(SystemCleanerWorker.java:62)
E/AndroidRuntime( 8271):        at eu.thedarken.sdm.c.b(AbstractWorker.java:178)
E/AndroidRuntime( 8271):        at eu.thedarken.sdm.c.i(AbstractWorker.java:189)
E/AndroidRuntime( 8271):        at eu.thedarken.sdm.oneclick.t.run(OneClickBox.java:136)
E/AndroidRuntime( 8271):        at java.lang.Thread.run(Thread.java:818)
W/ActivityManager( 2087):   Force finishing activity 1 eu.thedarken.sdm/.SDMMainActivity
```

**Comment by Geocfu** [comment]
Great. Waiting for the fix :)

**Comment by d4rken** [comment]
Should be fixed in v4.0.9
